secant/Features/BackupFlow/RecoveryPhraseValidationStep.swift

```diff
@@ -0,0 +1,145 @@
+                completion: [RecoveryPhraseStepCompletion(groupIndex: group, word: word)],
+                missingWordsChips: newMissingWords
+            )
+
+        case let .incomplete(phrase, missingIndices, completion, missingWordsChips):
+            guard let missingChipIndex = missingWordsChips.firstIndex(of: chip) else { return step }
+
+            if completion.count < (RecoveryPhraseValidationState.phraseChunks - 1) {
+                var newMissingWords = missingWordsChips
+                newMissingWords[missingChipIndex] = .empty
+
+                var newCompletionState = Array(completion)
+                newCompletionState.append(RecoveryPhraseStepCompletion(groupIndex: group, word: word))
+
+                return .incomplete(
+                    phrase: phrase,
+                    missingIndices: missingIndices,
+                    completion: newCompletionState,
+                    missingWordsChips: newMissingWords
+                )
+            } else {
+                var newCompletion = completion
+                newCompletion.append(RecoveryPhraseStepCompletion(groupIndex: group, word: word))
+
+                return .complete(
+                    phrase: phrase,
+                    missingIndices: missingIndices,
+                    completion: newCompletion,
+                    missingWordsChips: missingWordsChips
+                )
+            }
+        default:
+            return step
+        }
+    }
+
+    static func resultingPhrase(
+        from completion: [RecoveryPhraseStepCompletion],
+        missingIndices: [Int],
+        originalPhrase: RecoveryPhrase,
+        numberOfGroups: Int
+    ) -> [String] {
+        precondition(missingIndices.count == completion.count)
+        precondition(completion.count == numberOfGroups)
+
+        var words = originalPhrase.words
+        let groupLength = words.count / numberOfGroups
+        // iterate based on the completions the user did on the UI
+        for wordCompletion in completion {
+            // figure out which phrase group (chunk) this completion belongs to
+            let i = wordCompletion.groupIndex
+            // validate that's the right number
+            precondition(i < numberOfGroups)
+            // get the missing index that the user did this completion for on the given group
+            let missingIndex = missingIndices[i]
+            // figure out what this means in terms of the whole recovery phrase
+            let concreteIndex = i * groupLength + missingIndex
+            precondition(concreteIndex < words.count)
+            // replace the word on the copy of the original phrase with the completion the user did
+            words[concreteIndex] = wordCompletion.word
+        }
+
+        return words
+    }
+    /**
+    validates that the resulting word on the complete state matches the original word and moves the state into either valid or invalid
+    */
+    mutating func validate() {
```

Comment:
In a struct based world, this would be removed completely and the phase of the state would be computed based on the current values, so the state would automatically transition to .valid when the underlying data was in the right shape.